src/database/sql/sql.go

```diff
@@ -1300,14 +1300,14 @@ func (db *DB) putConn(dc *driverConn, err error, resetSession bool) {
 // If err == nil, then dc must not equal nil.
 // If a connRequest was fulfilled or the *driverConn was placed in the
 // freeConn list, then true is returned, otherwise false is returned.
-func (db *DB) putConnDBLocked(dc *driverConn, err error) bool {
+func (db *DB) putConnDBLocked(dc *driverConn, err error, newConn bool) bool {
 	if db.closed {
 		return false
 	}
 	if db.maxOpen > 0 && db.numOpen > db.maxOpen {
 		return false
 	}
-	if c := len(db.connRequests); c > 0 {
+	if c := len(db.connRequests); c > 0 && newConn {
```

Comment:
is it possible we hit this case, append the connection to the freeconn list, and then get stuck because there is an open connection but also somebody is writing on a fresh new connection?

Reply:
slash can we write a test for this case?

Reply:
It turns out there is a test that will already fail with a deadlock because of this: https://github.com/samsarahq/golang-go/blob/master/src/database/sql/sql_test.go#L1876

I wasn't seeing the failure due to test caching.